src/metabase/api/card.clj

```diff
@@ -71,52 +167,68 @@
+  {name                   NonEmptyString
+   public_perms           PublicPerms
+   display                NonEmptyString
+   visualization_settings Dict
+   archived               Boolean}
   (write-check Card id)
   (upd-non-nil-keys Card id
-                    :dataset_query dataset_query
-                    :description description
-                    :display display
-                    :name name
-                    :public_perms public_perms
-                    :visualization_settings visualization_settings)
-  (events/publish-event :card-update (assoc (sel :one Card :id id) :actor_id *current-user-id*)))
+    :dataset_query          dataset_query
+    :description            description
+    :display                display
+    :name                   name
+    :public_perms           public_perms
+    :visualization_settings visualization_settings
+    :archived               archived)
+  (events/publish-event :card-update (assoc (Card id) :actor_id *current-user-id*)))
 
 (defendpoint DELETE "/:id"
   "Delete a `Card`."
   [id]
   (write-check Card id)
   (let-404 [card (sel :one Card :id id)]
     (write-check card)
-    (let [result (cascade-delete Card :id id)]
-      (events/publish-event :card-delete (assoc card :actor_id *current-user-id*))
-      result)))
+    (u/prog1 (cascade-delete Card :id id)
+      (events/publish-event :card-delete (assoc card :actor_id *current-user-id*)))))
 
 (defendpoint GET "/:id/favorite"
   "Has current user favorited this `Card`?"
   [id]
-  {:favorite (boolean (some->> *current-user-id*
-                               (exists? CardFavorite :card_id id :owner_id)))})
+  {:favorite (boolean (when *current-user-id*
+                        (exists? CardFavorite :card_id, id :owner_id *current-user-id*)))})
 
 (defendpoint POST "/:card-id/favorite"
   "Favorite a Card."
   [card-id]
-  (ins CardFavorite :card_id card-id :owner_id *current-user-id*))
+  (ins CardFavorite :card_id card-id, :owner_id *current-user-id*))
 
 (defendpoint DELETE "/:card-id/favorite"
   "Unfavorite a Card."
   [card-id]
   (let-404 [id (sel :one :id CardFavorite :card_id card-id, :owner_id *current-user-id*)]
     (del CardFavorite :id id)))
 
+(defendpoint POST "/:card-id/labels"
+  "Update the set of `Labels` that apply to a `Card`."
+  [card-id :as {{:keys [label_ids]} :body}]
+  {label_ids [Required ArrayOfIntegers]}
+  (write-check Card card-id)
+  (let [[labels-to-remove labels-to-add] (data/diff (set (sel :many :field [CardLabel :label_id] :card_id card-id))
+                                                    (set label_ids))]
+    (doseq [label-id labels-to-remove]
+      (cascade-delete CardLabel :label_id label-id, :card_id card-id))
+    (doseq [label-id labels-to-add]
+      (ins CardLabel :label_id label-id, :card_id card-id)))
+  ;; TODO - Should this endpoint return something more useful instead ?
+  {:status :ok})
```

Comment:
@camsaul it should return the final set of labels which are now persisted against the card.